### DISCUSSION TITLE
Pylance should have implicit includes for Notebooks. So that, e.g. IPython display function is recognized

### DISCUSSION THREAD

**Discussion by claudiaregio** [microsoft/pylance-release#2713] [Enhancement]
-   Language Server version: 2021.7.5
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): Conda 4.10.3

I'm not sure why the following below is showing up as an error, get the same error in a python file

![image](https://user-images.githubusercontent.com/19276747/126908735-2f22b515-29a2-40a7-ba71-530880efcfd3.png)

![image](https://user-images.githubusercontent.com/19276747/126908826-d87a5112-67e8-4d13-a407-823953400071.png)

**Comment by erictraut** [comment]
`display` is not a symbol that is in the builtins namespace, and you are not importing it into your module's global namespace.

Does `ipywidgets` inject new symbols into the builtins or module namespace even if you don't import them? If so, pylance would have no idea that it's doing under the covers. This is a really terrible practice that some older Python libraries did. If you explicitly import the symbol into your module's namespace, then it will work as expected.

Alternatively, you can disable the `reportUndefinedVariable` diagnostic check for your project.

**Comment by jakebailey** [comment]
This symbol is specific to IPython notebooks, not the widget library, I believe. From the IPython docs:

> Since IPython 5.4 and 6.1 display() is automatically made available to the user without import. If you are using display in a document that might be used in a pure python context or with older version of IPython, use the following import at the top of your file:
>
> ```python
> from IPython.display import display
> ```

Normally we'd close issues for custom globals like this, but since Python notebooks are something we'd like to fully support, I'm going to leave this open for now to gather feedback and to think about how we might handle this (if at all).

**Comment by claudiaregio** [comment]
Ah okay I didn't know that. It would have been nice if there was a way for me to resolve it or ignore somehow for the time being. Since the code works I'd hope I don't have to add more code or change a setting to resolve a "false" diagnostic.

**Comment by jakebailey** [comment]
You can throw `# type: ignore` at the end of the line to ignore the diagnostic, or disable the `reportUndefinedVariable` diagnostic with a severity override.

Or, you can throw `from IPython.display import display` at the top of your file, which will get the actual `display` function (and probably some more useful info).

**Comment by shamrin** [comment]
I've reported same issue in vscode-jupyter project: https://github.com/microsoft/vscode-jupyter/issues/8539

**Comment by rchiodo** [comment]
Our concat document can likely handle this. Not sure yet if we should or not. Essentially we'd put this at the top of the concat doc:

```
from IPython.display import display
```
and then pylance would understand ```display```.

**Comment by heejaechang** [comment]
@rchiodo is it only display or "from IPython.display import *" ?

**Comment by rchiodo** [comment]
> @rchiodo is it only display or "from IPython.display import *" ?

Second one I think? That would enable stuff like this:

```
from IPython.display import *

display_javascript()
```

**Comment by DavidHopkinsFbr** [comment]
I've tried to use `from IPython.display import display` and `from IPython.display import *`, but it's not resolving the issue for me. I now get an `Import "IPython.display" could not be resolved` error from Pylance _as well as_ the pre-existing `"display" is not defined` errors.
![image](https://user-images.githubusercontent.com/49743077/158913730-21c7db29-d3a0-47d4-b146-1864e823984e.png)

My Jupyter notebook uses an Anaconda environment, which might be relevant.

**Comment by rchiodo** [comment]
Sorry @DavidHopkinsFbr the comment I made was in reference to the internals of pylance and what it should do. Not what a user should add to resolve the problem. The root cause of the issue is that pylance doesn't have the implicit includes that jupyter includes.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by jeisonmp** [comment]
I am yet with this problem. Was this solved?

**Reply by heejaechang** [reply]
this should have been fixed. make sure to use the latest python, jupyter, pylance extensions.

![image](https://user-images.githubusercontent.com/1333179/222985525-023f2ab8-4adc-4413-8a18-bf1063fc7a81.png)

**Reply by philipnye** [reply]
Still present for me using Python v2023.6.0, Jupyter v2023.3.1000892223, Pylance v2023.3.40, VSCode 1.77.0.

**Reply by heejaechang** [reply]
Can you provide us some logs as described in [the troubleshooting guide](https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#filing-an-issue)?

we need a way to repro, otherwise, there is not much we can do.

**Reply by rchiodo** [reply]
This might be a problem with the kernel you have selected not having `IPython` installed in it yet. If you run a cell, the warning should go away.

**Reply by philipnye** [reply]
Please find my language server log attached and a minimal code example (saved as .txt as .py can't be uploaded). I'm running code using Jupyter code cells within my .py scripts rather than Jupyter notebooks. 

@rchiodo The warning doesn't go away when I run the cell.

Let me know if there's anything else I can provide.

[issue_example.txt](https://github.com/microsoft/pylance-release/files/11147083/issue_example.txt)
[language_server_log.txt](https://github.com/microsoft/pylance-release/files/11147084/language_server_log.txt)

**Reply by rchiodo** [reply]
>  I'm running code using Jupyter code cells within my .py scripts rather than Jupyter notebooks.

The display should be a warning in this case. The hidden imports are only added for notebooks. Or are you seeing it in the interactive window?

**Reply by heejaechang** [reply]
by the way, can you open an issue rather than using discussion so that we can track the issue properly?

**Reply by heejaechang** [reply]
>  I'm running code using Jupyter code cells within my .py scripts rather than Jupyter notebooks

what do you mean by that? are you talking about `interactive window`?

**Reply by heejaechang** [reply]
ignore what I said above, I figured out what you meant.

**Reply by heejaechang** [reply]
as @rchiodo said, it is due to us only resolving ipython display on notebook (https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/analyzer/sourceFile.ts#L1529)

I am not sure whether we should implicitly import ipython on regular py file which uses Jupyter cell feature or not.

for now, you need to explicitly import display by doing `from IPython.display import display`

if you open an issue, people could discuss.
